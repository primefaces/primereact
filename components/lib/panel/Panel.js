import * as React from 'react';
import { CSSTransition } from '../csstransition/CSSTransition';
import { useMountEffect } from '../hooks/Hooks';
import { Ripple } from '../ripple/Ripple';
import { classNames, IconUtils, ObjectUtils, UniqueComponentId } from '../utils/Utils';
import { PanelBase } from './PanelBase';

export const Panel = React.forwardRef((inProps, ref) => {
    const props = PanelBase.getProps(inProps);

    const [idState, setIdState] = React.useState(props.id);
    const [collapsedState, setCollapsedState] = React.useState(props.collapsed);
    const elementRef = React.useRef(ref);
    const contentRef = React.useRef(null);
    const collapsed = props.toggleable ? (props.onToggle ? props.collapsed : collapsedState) : false;
    const headerId = idState + '_header';
    const contentId = idState + '_content';

    const toggle = (event) => {
        if (props.toggleable) {
            collapsed ? expand(event) : collapse(event);

            if (props.onToggle) {
                props.onToggle({
                    originalEvent: event,
                    value: !collapsed
                });
            }
        }

        event.preventDefault();
    };

    const expand = (event) => {
        if (!props.onToggle) {
            setCollapsedState(false);
        }

        props.onExpand && props.onExpand(event);
    };

    const collapse = (event) => {
        if (!props.onToggle) {
            setCollapsedState(true);
        }

        props.onCollapse && props.onCollapse(event);
    };

    React.useImperativeHandle(ref, () => ({
        props,
        getElement: () => elementRef.current,
        getContent: () => contentRef.current
    }));

    React.useEffect(() => {
        ObjectUtils.combinedRefs(elementRef, ref);
    }, [elementRef, ref]);

    useMountEffect(() => {
        if (!idState) {
            setIdState(UniqueComponentId());
        }
    });

    const createToggleIcon = () => {
        if (props.toggleable) {
            const buttonId = idState + '_label';
            const toggleIcon = collapsed ? props.expandIcon : props.collapseIcon;

            return (
                <button className="p-panel-header-icon p-panel-toggler p-link" onClick={toggle} id={buttonId} aria-controls={contentId} aria-expanded={!collapsed} role="tab">
                    {IconUtils.getJSXIcon(toggleIcon, undefined, { props, collapsed })}
                    <Ripple />
                </button>
            );
        }

        return null;
    };

    const createHeader = () => {
        const header = ObjectUtils.getJSXElement(props.header, props);
        const icons = ObjectUtils.getJSXElement(props.icons, props);
        const togglerElement = createToggleIcon();
        const titleElement = (
            <span className="p-panel-title" id={headerId}>
                {header}
            </span>
        );
        const iconsElement = (
            <div className="p-panel-icons">
                {icons}
                {togglerElement}
            </div>
        );
        const content = (
            <div className="p-panel-header">
                {titleElement}
                {iconsElement}
            </div>
        );

        if (props.headerTemplate) {
            const defaultContentOptions = {
                className: 'p-panel-header',
                titleClassName: 'p-panel-title',
                iconsClassName: 'p-panel-icons',
                togglerClassName: 'p-panel-header-icon p-panel-toggler p-link',
                togglerIconClassName: collapsed ? props.expandIcon : props.collapseIcon,
                onTogglerClick: toggle,
                titleElement,
                iconsElement,
                togglerElement,
                element: content,
                props,
                collapsed
            };

            return ObjectUtils.getJSXElement(props.headerTemplate, defaultContentOptions);
        } else if (props.header || props.toggleable) {
            return content;
        }

        return null;
    };

    const createContent = () => {
        return (
            <CSSTransition nodeRef={contentRef} classNames="p-toggleable-content" timeout={{ enter: 1000, exit: 450 }} in={!collapsed} unmountOnExit options={props.transitionOptions}>
                <div ref={contentRef} className="p-toggleable-content" aria-hidden={collapsed} role="region" id={contentId} aria-labelledby={headerId}>
                    <div className="p-panel-content">{props.children}</div>
                </div>
            </CSSTransition>
        );
    };

    const otherProps = PanelBase.getOtherProps(props);
    const className = classNames(
        'p-panel p-component',
        {
            'p-panel-toggleable': props.toggleable
        },
        props.className
    );
    const header = createHeader();
    const content = createContent();

    return (
        <div id={props.id} ref={elementRef} className={className} style={props.style} {...otherProps}>
            {header}
            {content}
        </div>
    );
});

Panel.displayName = 'Panel';
