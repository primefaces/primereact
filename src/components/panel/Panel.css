.p-panel-header {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.p-panel-title {
    line-height: 1;
}

.p-panel-header-icon {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    text-decoration: none;
    overflow: hidden;
    position: relative;
}
