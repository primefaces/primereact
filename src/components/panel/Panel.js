import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { CSSTransition } from '../transition/CSSTransition';
import { classNames } from '../utils/ClassNames';
import UniqueComponentId from '../utils/UniqueComponentId';
import ObjectUtils from '../utils/ObjectUtils';
import { Ripple } from '../ripple/Ripple';

export class Panel extends Component {

    static defaultProps = {
        id: null,
        header: null,
        headerTemplate: null,
        toggleable: null,
        style: null,
        className: null,
        collapsed: null,
        expandIcon: 'pi pi-plus',
        collapseIcon: 'pi pi-minus',
        icons: null,
        transitionOptions: null,
        onExpand: null,
        onCollapse: null,
        onToggle: null
    }

    static propTypes = {
        id: PropTypes.string,
        header: PropTypes.any,
        headerTemplate: PropTypes.any,
        toggleable: PropTypes.bool,
        style: PropTypes.object,
        className: PropTypes.string,
        collapsed: PropTypes.bool,
        expandIcon: PropTypes.string,
        collapseIcon: PropTypes.string,
        icons: PropTypes.any,
        transitionOptions: PropTypes.object,
        onExpand: PropTypes.func,
        onCollapse: PropTypes.func,
        onToggle: PropTypes.func
    };

    constructor(props) {
        super(props);
        let state = {
            id: this.props.id
        };

        if (!this.props.onToggle) {
            state = {
                ...state,
                collapsed: this.props.collapsed
            };
        }

        this.state = state;

        this.toggle = this.toggle.bind(this);
        this.contentRef = React.createRef();
    }

    toggle(event) {
        if (this.props.toggleable) {
            const collapsed = this.props.onToggle ? this.props.collapsed : this.state.collapsed;

            if (collapsed)
                this.expand(event);
            else
                this.collapse(event);

            if (this.props.onToggle) {
                this.props.onToggle({
                    originalEvent: event,
                    value: !collapsed
                });
            }
        }

        event.preventDefault();
    }

    expand(event) {
        if (!this.props.onToggle) {
            this.setState({ collapsed: false });
        }

        if (this.props.onExpand) {
            this.props.onExpand(event);
        }
    }

    collapse(event) {
        if (!this.props.onToggle) {
            this.setState({ collapsed: true });
        }

        if (this.props.onCollapse) {
            this.props.onCollapse(event);
        }
    }

    isCollapsed() {
        return this.props.toggleable ? (this.props.onToggle ? this.props.collapsed : this.state.collapsed) : false;
    }

    componentDidMount() {
        if (!this.state.id) {
            this.setState({ id: UniqueComponentId() });
        }
    }

    renderToggleIcon(collapsed) {
        if (this.props.toggleable) {
            const id = this.state.id + '_label';
            const ariaControls = this.state.id + '_content';
            const toggleIcon = collapsed ? this.props.expandIcon : this.props.collapseIcon;

            return (
                <button className="p-panel-header-icon p-panel-toggler p-link" onClick={this.toggle}
                    id={id} aria-controls={ariaControls} aria-expanded={!collapsed} role="tab">
                    <span className={toggleIcon}></span>
                    <Ripple />
                </button>
            );
        }

        return null;
    }

    renderHeader(collapsed) {
        const header = ObjectUtils.getJSXElement(this.props.header, this.props);
        const icons = ObjectUtils.getJSXElement(this.props.icons, this.props);
        const togglerElement = this.renderToggleIcon(collapsed);
        const titleElement = <span className="p-panel-title" id={this.state.id + '_header'}>{header}</span>;
        const iconsElement = (
            <div className="p-panel-icons">
                {icons}
                {togglerElement}
            </div>
        );
        const content = (
            <div className="p-panel-header">
                {titleElement}
                {iconsElement}
            </div>
        );

        if (this.props.headerTemplate) {
            const defaultContentOptions = {
                className: 'p-panel-header',
                titleClassName: 'p-panel-title',
                iconsClassName: 'p-panel-icons',
                togglerClassName: 'p-panel-header-icon p-panel-toggler p-link',
                togglerIconClassName: collapsed ? this.props.expandIcon : this.props.collapseIcon,
                onTogglerClick: this.toggle,
                titleElement,
                iconsElement,
                togglerElement,
                element: content,
                props: this.props,
                collapsed
            };

            return ObjectUtils.getJSXElement(this.props.headerTemplate, defaultContentOptions);
        }
        else if (this.props.header || this.props.toggleable) {
            return content;
        }

        return null;
    }

    renderContent(collapsed) {
        const id = this.state.id + '_content';

        return (
            <CSSTransition nodeRef={this.contentRef} classNames="p-toggleable-content" timeout={{ enter: 1000, exit: 450 }} in={!collapsed} unmountOnExit options={this.props.transitionOptions}>
                <div ref={this.contentRef} className="p-toggleable-content" aria-hidden={collapsed} role="region" id={id} aria-labelledby={this.state.id + '_header'}>
                    <div className="p-panel-content">
                        {this.props.children}
                    </div>
                </div>
            </CSSTransition>
        );
    }

    render() {
        const className = classNames('p-panel p-component', { 'p-panel-toggleable': this.props.toggleable }, this.props.className);
        const collapsed = this.isCollapsed();
        const header = this.renderHeader(collapsed);
        const content = this.renderContent(collapsed);

        return (
            <div id={this.props.id} className={className} style={this.props.style}>
                {header}
                {content}
            </div>
        );
    }
}
