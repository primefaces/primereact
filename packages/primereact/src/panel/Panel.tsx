'use client';
import { Component, withComponent } from '@primereact/core/component';
import { usePanel } from '@primereact/headless/panel';
import { styles } from '@primereact/styles/panel';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { defaultProps } from './Panel.props';
import { PanelCollapse } from './collapse';
import { PanelContent } from './content';
import { PanelFooter } from './footer';
import { PanelHeader } from './header';
import { PanelHeaderActions } from './headeractions';
import { PanelTitle } from './title';

export const Panel = withComponent({
    name: 'Panel',
    defaultProps,
    styles,
    setup: (instance) => {
        const panel = usePanel(instance?.inProps);

        return panel;
    },
    render: (instance) => {
        const {
            id,
            props,
            ptmi,
            cx,
            // element refs
            elementRef
        } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root')
            },
            ptmi('root')
        );

        return <Component as={props.as} {...rootProps} ref={elementRef} children={props.children} />;
    },
    components: {
        Header: PanelHeader,
        Title: PanelTitle,
        HeaderActions: PanelHeaderActions,
        Content: PanelContent,
        Footer: PanelFooter,
        Collapse: PanelCollapse
    }
});
