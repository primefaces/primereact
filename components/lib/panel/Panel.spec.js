import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { snapshot } from '../../test';
import { Panel } from './Panel';

const template = (options) => {
    const toggleIcon = options.collapsed ? 'pi pi-chevron-down' : 'pi pi-chevron-up';
    const className = `${options.className} justify-content-start`;
    const titleClassName = `${options.titleClassName} pl-1`;

    return (
        <div className={className}>
            <button className={options.togglerClassName} onClick={options.onTogglerClick}>
                <span className={toggleIcon}></span>
            </button>
            <span className={titleClassName}>Header</span>
        </div>
    );
};

describe('Panel', () => {
    snapshot(<Panel />, 'default');
    snapshot(
        <Panel id="Panel" header="Simple Panel">
            Content
        </Panel>,
        'header'
    );
    snapshot(
        <Panel headerTemplate={template} toggleable>
            Content
        </Panel>,
        'headerTemplate'
    );
    test('when Panel is toggleable it will toggle when clicked', async () => {
        // Arrange
        const toggleOn = jest.fn();
        const { container } = render(
            <Panel header="Toggleable" toggleable onToggle={toggleOn}>
                <p>Lorem ipsum dolor sit amet</p>
            </Panel>
        );
        const toggler = container.getElementsByClassName('p-panel-toggler')[0];

        expect(container).toMatchSnapshot('toggleable-open');

        // Act
        await userEvent.click(toggler);

        // Assert
        expect(toggleOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('toggleable-closed');
    });
    test('when Panel is toggleable it must expand and collapse', async () => {
        // Arrange
        const expandOn = jest.fn();
        const collapseOn = jest.fn();
        const { container } = render(
            <Panel header="Expand/Collapse" toggleable onExpand={expandOn} onCollapse={collapseOn}>
                <p>Lorem ipsum dolor sit amet</p>
            </Panel>
        );
        const toggler = container.getElementsByClassName('p-panel-toggler')[0];

        expect(container).toMatchSnapshot('expandable-open');

        // Act
        await userEvent.click(toggler);

        // Assert
        expect(expandOn).toHaveBeenCalledTimes(0);
        expect(collapseOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('expandable-closed');

        // Act
        await userEvent.click(toggler);

        // Assert
        expect(expandOn).toHaveBeenCalledTimes(1);
        expect(collapseOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('expandable-open');
    });
});
