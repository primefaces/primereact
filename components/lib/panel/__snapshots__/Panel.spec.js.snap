// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel default 1`] = `
<div>
  <div
    class="p-panel p-component"
  >
    <div
      aria-hidden="false"
      aria-labelledby="pr_id_1_header"
      class="p-toggleable-content"
      id="pr_id_1_content"
      role="region"
    >
      <div
        class="p-panel-content"
      />
    </div>
  </div>
</div>
`;

exports[`Panel header 1`] = `
<div>
  <div
    class="p-panel p-component"
    id="Panel"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="Panel_header"
      >
        Simple Panel
      </span>
      <div
        class="p-panel-icons"
      />
    </div>
    <div
      aria-hidden="false"
      aria-labelledby="Panel_header"
      class="p-toggleable-content"
      id="Panel_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel headerTemplate 1`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header justify-content-start"
    >
      <button
        class="p-panel-header-icon p-panel-toggler p-link"
      >
        <span
          class="pi pi-chevron-up"
        />
      </button>
      <span
        class="p-panel-title pl-1"
      >
        Header
      </span>
    </div>
    <div
      aria-labelledby="pr_id_2_header"
      class="p-toggleable-content"
      id="pr_id_2_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel when Panel is toggleable it must expand and collapse: expandable-closed 1`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="pr_id_4_header"
      >
        Expand/Collapse
      </span>
      <div
        class="p-panel-icons"
      >
        <button
          aria-controls="pr_id_4_content"
          aria-expanded="false"
          class="p-panel-header-icon p-panel-toggler p-link"
          id="pr_id_4_label"
          role="tab"
        >
          <span
            class="pi pi-plus"
          />
        </button>
      </div>
    </div>
    <div
      aria-hidden="true"
      aria-labelledby="pr_id_4_header"
      class="p-toggleable-content p-toggleable-content-exit p-toggleable-content-exit-active"
      id="pr_id_4_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel when Panel is toggleable it must expand and collapse: expandable-open 1`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="pr_id_4_header"
      >
        Expand/Collapse
      </span>
      <div
        class="p-panel-icons"
      >
        <button
          aria-controls="pr_id_4_content"
          aria-expanded="true"
          class="p-panel-header-icon p-panel-toggler p-link"
          id="pr_id_4_label"
          role="tab"
        >
          <span
            class="pi pi-minus"
          />
        </button>
      </div>
    </div>
    <div
      aria-labelledby="pr_id_4_header"
      class="p-toggleable-content"
      id="pr_id_4_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel when Panel is toggleable it must expand and collapse: expandable-open 2`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="pr_id_4_header"
      >
        Expand/Collapse
      </span>
      <div
        class="p-panel-icons"
      >
        <button
          aria-controls="pr_id_4_content"
          aria-expanded="true"
          class="p-panel-header-icon p-panel-toggler p-link"
          id="pr_id_4_label"
          role="tab"
        >
          <span
            class="pi pi-minus"
          />
        </button>
      </div>
    </div>
    <div
      aria-hidden="false"
      aria-labelledby="pr_id_4_header"
      class="p-toggleable-content p-toggleable-content-enter p-toggleable-content-enter-active"
      id="pr_id_4_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel when Panel is toggleable it will toggle when clicked: toggleable-closed 1`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="pr_id_3_header"
      >
        Toggleable
      </span>
      <div
        class="p-panel-icons"
      >
        <button
          aria-controls="pr_id_3_content"
          aria-expanded="true"
          class="p-panel-header-icon p-panel-toggler p-link"
          id="pr_id_3_label"
          role="tab"
        >
          <span
            class="pi pi-minus"
          />
        </button>
      </div>
    </div>
    <div
      aria-labelledby="pr_id_3_header"
      class="p-toggleable-content"
      id="pr_id_3_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Panel when Panel is toggleable it will toggle when clicked: toggleable-open 1`] = `
<div>
  <div
    class="p-panel p-component p-panel-toggleable"
  >
    <div
      class="p-panel-header"
    >
      <span
        class="p-panel-title"
        id="pr_id_3_header"
      >
        Toggleable
      </span>
      <div
        class="p-panel-icons"
      >
        <button
          aria-controls="pr_id_3_content"
          aria-expanded="true"
          class="p-panel-header-icon p-panel-toggler p-link"
          id="pr_id_3_label"
          role="tab"
        >
          <span
            class="pi pi-minus"
          />
        </button>
      </div>
    </div>
    <div
      aria-labelledby="pr_id_3_header"
      class="p-toggleable-content"
      id="pr_id_3_content"
      role="region"
    >
      <div
        class="p-panel-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;
